{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "57c14f3e-f6c0-4307-923b-cca4475b01c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "This file is used to prepare transcriptomic data for running in DynGENIE3 to build a Gene Regulatory Network\n",
    "\n",
    "KEGG terms are used to get protein info / transcription factors\n",
    "Transcriptomic read counts are used from Daphnia magna\n",
    "\n",
    "1. Put timepoints in rows and genes in columns\n",
    "2. Make timepoints numeric\n",
    "3. Use KEGG file to match KEGG orthologs (transcription factors) with genes present in the transcriptomic data file\n",
    "4. Extract gene order to maintain the correct order for alphas in the next file\n",
    "\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6fd6422-3f11-4e3c-af9e-2977c1b2e1ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "1fdc3ba9-1cd7-46c4-8269-324b30622b1c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/v1/ymptnwwx3x1b7_mw0nns09wr0000gn/T/ipykernel_6261/1602555744.py:1: DtypeWarning: Columns (2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119,120,121,122,123,124,125,126,127,128,129,130,131,132,133,134,135,136,137,138,139,140,141,142,143,144,145,146,147,148,149,150,151,152,153,154,155,156,157,158,159,160) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  genes = pd.read_csv('../../omics/rna_vst_proc.csv')\n"
     ]
    }
   ],
   "source": [
    "genes = pd.read_csv('../../omics/rna_vst_proc.csv')\n",
    "tfs = pd.read_csv('../../ortholog/protein_info_kegg_terms.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "27dd8d54-45ea-4724-b5a4-cbdbaf503899",
   "metadata": {},
   "outputs": [],
   "source": [
    "#delete 2nd row (concatted treatment + timepoint info)\n",
    "genes_dropped = genes.drop(1)\n",
    "\n",
    "#separate diff conditions into 3 files, control, low, high\n",
    "genes_list = genes_dropped['treatment'].tolist()\n",
    "\n",
    "#function to make timepoints rows and genes columns\n",
    "def format_GRN(df,string_name):\n",
    "\n",
    "    #find condition information (control, low, high dose)\n",
    "    control_columns = [col for col in genes_dropped.columns if col.startswith(string_name)]\n",
    "    \n",
    "    #separate conditions\n",
    "    df = genes_dropped[control_columns]\n",
    "\n",
    "    \n",
    "    #add gene names\n",
    "    df['time_points'] = genes_list\n",
    "    df = df[['time_points'] + control_columns]\n",
    "\n",
    "    #get list of timepoints\n",
    "    time_points = list(df.iloc[0])\n",
    "    timepoints = time_points[1:]\n",
    "    #delete timepoints row\n",
    "    df = df.iloc[1:]\n",
    "\n",
    "    # Rename the columns to timepoints and gene names\n",
    "    df.columns = ['time_points'] + timepoints\n",
    "\n",
    "    #change to genes as column names\n",
    "    df_transposed = df.T.reset_index()  # Reset the index to make 'time_points' a column\n",
    "\n",
    "    # Set the new column names\n",
    "    df_transposed.columns = df_transposed.iloc[0]\n",
    "\n",
    "    # Drop the first row\n",
    "    df_transposed = df_transposed.drop(df_transposed.index[0])\n",
    "    df_transposed['time_points'] = timepoints\n",
    "\n",
    "    #get mean for each timepoint experiment\n",
    "    grouped_df = df_transposed.groupby('time_points').mean()\n",
    "    #reset index and add timepoints as a column\n",
    "    grouped_times = grouped_df.index.tolist()\n",
    "    grouped_df.reset_index(drop=True, inplace=True)\n",
    "    grouped_df.insert(0,'time_points',grouped_times)\n",
    "\n",
    "    \n",
    "    return grouped_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "23c69ce9-e23d-4477-8473-f32a0d1e7a24",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/v1/ymptnwwx3x1b7_mw0nns09wr0000gn/T/ipykernel_6261/1380820900.py:18: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df['time_points'] = genes_list\n",
      "/var/folders/v1/ymptnwwx3x1b7_mw0nns09wr0000gn/T/ipykernel_6261/1380820900.py:41: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.\n",
      "  grouped_df = df_transposed.groupby('time_points').mean()\n",
      "/var/folders/v1/ymptnwwx3x1b7_mw0nns09wr0000gn/T/ipykernel_6261/1380820900.py:45: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  grouped_df.insert(0,'time_points',grouped_times)\n",
      "/var/folders/v1/ymptnwwx3x1b7_mw0nns09wr0000gn/T/ipykernel_6261/1380820900.py:18: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df['time_points'] = genes_list\n",
      "/var/folders/v1/ymptnwwx3x1b7_mw0nns09wr0000gn/T/ipykernel_6261/1380820900.py:41: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.\n",
      "  grouped_df = df_transposed.groupby('time_points').mean()\n",
      "/var/folders/v1/ymptnwwx3x1b7_mw0nns09wr0000gn/T/ipykernel_6261/1380820900.py:45: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  grouped_df.insert(0,'time_points',grouped_times)\n",
      "/var/folders/v1/ymptnwwx3x1b7_mw0nns09wr0000gn/T/ipykernel_6261/1380820900.py:18: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df['time_points'] = genes_list\n",
      "/var/folders/v1/ymptnwwx3x1b7_mw0nns09wr0000gn/T/ipykernel_6261/1380820900.py:41: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.\n",
      "  grouped_df = df_transposed.groupby('time_points').mean()\n",
      "/var/folders/v1/ymptnwwx3x1b7_mw0nns09wr0000gn/T/ipykernel_6261/1380820900.py:45: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  grouped_df.insert(0,'time_points',grouped_times)\n"
     ]
    }
   ],
   "source": [
    "control_df = format_GRN('control_df', 'control')\n",
    "low_df = format_GRN('low_df', 'low')\n",
    "high_df = format_GRN('high_df', 'high')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2e959522-aef2-4b17-9085-cae9ca4c5f19",
   "metadata": {},
   "outputs": [],
   "source": [
    "#make time_points numeric\n",
    "def make_numeric(df):\n",
    "    times = df['time_points'].tolist()\n",
    "    new_times = []\n",
    "    for point in times:\n",
    "        if point == '12H':\n",
    "            new_times.append(12)\n",
    "        elif point == '1H':\n",
    "            new_times.append(1)\n",
    "        elif point == '24H':\n",
    "            new_times.append(24)\n",
    "        elif point == '2H':\n",
    "            new_times.append(2)\n",
    "        elif point == '4D':\n",
    "            new_times.append(4*24)\n",
    "        elif point == '5D':\n",
    "            new_times.append(5*24)\n",
    "        elif point == '6D':\n",
    "            new_times.append(6*24)\n",
    "        elif point == '6H':\n",
    "            new_times.append(6)\n",
    "        elif point == '7D':\n",
    "            new_times.append(7*24)\n",
    "        \n",
    "    df['time_points'] = new_times\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "940bd877-c87e-4fad-bff9-78c55c766e58",
   "metadata": {},
   "outputs": [],
   "source": [
    "control_df = make_numeric(control_df)\n",
    "low_df = make_numeric(low_df)\n",
    "high_df = make_numeric(high_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "d407439f-5fe4-4bc1-8bb3-61e809111ca5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#get a file of transcription factors which exist in our dataset\n",
    "genes_list = control_df.columns.tolist()\n",
    "tfs_genes = tfs.loc[tfs['regulatory'].isin(genes_list)]\n",
    "tfs_genes.to_csv('regulatory_genes.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "08d6a672-f23a-42c1-8600-abd9d29258e0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>time_points</th>\n",
       "      <th>Dapma7bEVm004153</th>\n",
       "      <th>Dapma7bEVm004154</th>\n",
       "      <th>Dapma7bEVm004155</th>\n",
       "      <th>Dapma7bEVm004156</th>\n",
       "      <th>Dapma7bEVm004157</th>\n",
       "      <th>Dapma7bEVm004158</th>\n",
       "      <th>Dapma7bEVm004159</th>\n",
       "      <th>Dapma7bEVm004160</th>\n",
       "      <th>Dapma7bEVm004161</th>\n",
       "      <th>...</th>\n",
       "      <th>Dapma7bEVm030739</th>\n",
       "      <th>Dapma7bEVm030740</th>\n",
       "      <th>Dapma7bEVm030744</th>\n",
       "      <th>Dapma7bEVm030748</th>\n",
       "      <th>Dapma7bEVm030749</th>\n",
       "      <th>Dapma7bEVm030753</th>\n",
       "      <th>Dapma7bEVm030754</th>\n",
       "      <th>Dapma7bEVm030755</th>\n",
       "      <th>Dapma7bEVm629765</th>\n",
       "      <th>Dapma7bEVm629948</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>12</td>\n",
       "      <td>7.352828</td>\n",
       "      <td>6.394134</td>\n",
       "      <td>6.776569</td>\n",
       "      <td>5.351553</td>\n",
       "      <td>9.928181</td>\n",
       "      <td>7.184691</td>\n",
       "      <td>6.473805</td>\n",
       "      <td>5.924651</td>\n",
       "      <td>5.790883</td>\n",
       "      <td>...</td>\n",
       "      <td>6.120184</td>\n",
       "      <td>4.629555</td>\n",
       "      <td>5.699719</td>\n",
       "      <td>5.862255</td>\n",
       "      <td>3.295922</td>\n",
       "      <td>4.994942</td>\n",
       "      <td>5.771966</td>\n",
       "      <td>5.693342</td>\n",
       "      <td>1.352579</td>\n",
       "      <td>4.201193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>7.642886</td>\n",
       "      <td>6.686304</td>\n",
       "      <td>6.997765</td>\n",
       "      <td>5.518121</td>\n",
       "      <td>9.323471</td>\n",
       "      <td>7.632350</td>\n",
       "      <td>7.082823</td>\n",
       "      <td>4.902429</td>\n",
       "      <td>5.643739</td>\n",
       "      <td>...</td>\n",
       "      <td>5.632045</td>\n",
       "      <td>4.970264</td>\n",
       "      <td>5.965698</td>\n",
       "      <td>5.408692</td>\n",
       "      <td>3.551104</td>\n",
       "      <td>5.315286</td>\n",
       "      <td>5.551543</td>\n",
       "      <td>6.029283</td>\n",
       "      <td>3.327923</td>\n",
       "      <td>4.118649</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>24</td>\n",
       "      <td>7.335762</td>\n",
       "      <td>6.658602</td>\n",
       "      <td>7.001900</td>\n",
       "      <td>6.626420</td>\n",
       "      <td>8.435655</td>\n",
       "      <td>7.160253</td>\n",
       "      <td>5.647166</td>\n",
       "      <td>4.818133</td>\n",
       "      <td>5.289874</td>\n",
       "      <td>...</td>\n",
       "      <td>5.851358</td>\n",
       "      <td>4.670966</td>\n",
       "      <td>5.671279</td>\n",
       "      <td>5.459678</td>\n",
       "      <td>4.014373</td>\n",
       "      <td>3.841160</td>\n",
       "      <td>4.700124</td>\n",
       "      <td>5.990468</td>\n",
       "      <td>1.844929</td>\n",
       "      <td>3.239676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>7.595695</td>\n",
       "      <td>6.476415</td>\n",
       "      <td>6.902540</td>\n",
       "      <td>4.623494</td>\n",
       "      <td>9.349699</td>\n",
       "      <td>6.509663</td>\n",
       "      <td>6.765066</td>\n",
       "      <td>6.486855</td>\n",
       "      <td>5.540381</td>\n",
       "      <td>...</td>\n",
       "      <td>4.645297</td>\n",
       "      <td>4.333281</td>\n",
       "      <td>5.045723</td>\n",
       "      <td>5.360963</td>\n",
       "      <td>3.749174</td>\n",
       "      <td>5.310375</td>\n",
       "      <td>5.690172</td>\n",
       "      <td>6.123093</td>\n",
       "      <td>3.083042</td>\n",
       "      <td>3.390129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>96</td>\n",
       "      <td>7.271276</td>\n",
       "      <td>6.710377</td>\n",
       "      <td>6.805541</td>\n",
       "      <td>4.317107</td>\n",
       "      <td>7.991139</td>\n",
       "      <td>4.674173</td>\n",
       "      <td>7.135154</td>\n",
       "      <td>7.128517</td>\n",
       "      <td>6.125496</td>\n",
       "      <td>...</td>\n",
       "      <td>5.832441</td>\n",
       "      <td>5.510325</td>\n",
       "      <td>6.079204</td>\n",
       "      <td>5.865779</td>\n",
       "      <td>3.142050</td>\n",
       "      <td>5.490820</td>\n",
       "      <td>5.715251</td>\n",
       "      <td>5.692676</td>\n",
       "      <td>1.804685</td>\n",
       "      <td>4.363281</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 11883 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "0  time_points  Dapma7bEVm004153  Dapma7bEVm004154  Dapma7bEVm004155  \\\n",
       "0           12          7.352828          6.394134          6.776569   \n",
       "1            1          7.642886          6.686304          6.997765   \n",
       "2           24          7.335762          6.658602          7.001900   \n",
       "3            2          7.595695          6.476415          6.902540   \n",
       "4           96          7.271276          6.710377          6.805541   \n",
       "\n",
       "0  Dapma7bEVm004156  Dapma7bEVm004157  Dapma7bEVm004158  Dapma7bEVm004159  \\\n",
       "0          5.351553          9.928181          7.184691          6.473805   \n",
       "1          5.518121          9.323471          7.632350          7.082823   \n",
       "2          6.626420          8.435655          7.160253          5.647166   \n",
       "3          4.623494          9.349699          6.509663          6.765066   \n",
       "4          4.317107          7.991139          4.674173          7.135154   \n",
       "\n",
       "0  Dapma7bEVm004160  Dapma7bEVm004161  ...  Dapma7bEVm030739  \\\n",
       "0          5.924651          5.790883  ...          6.120184   \n",
       "1          4.902429          5.643739  ...          5.632045   \n",
       "2          4.818133          5.289874  ...          5.851358   \n",
       "3          6.486855          5.540381  ...          4.645297   \n",
       "4          7.128517          6.125496  ...          5.832441   \n",
       "\n",
       "0  Dapma7bEVm030740  Dapma7bEVm030744  Dapma7bEVm030748  Dapma7bEVm030749  \\\n",
       "0          4.629555          5.699719          5.862255          3.295922   \n",
       "1          4.970264          5.965698          5.408692          3.551104   \n",
       "2          4.670966          5.671279          5.459678          4.014373   \n",
       "3          4.333281          5.045723          5.360963          3.749174   \n",
       "4          5.510325          6.079204          5.865779          3.142050   \n",
       "\n",
       "0  Dapma7bEVm030753  Dapma7bEVm030754  Dapma7bEVm030755  Dapma7bEVm629765  \\\n",
       "0          4.994942          5.771966          5.693342          1.352579   \n",
       "1          5.315286          5.551543          6.029283          3.327923   \n",
       "2          3.841160          4.700124          5.990468          1.844929   \n",
       "3          5.310375          5.690172          6.123093          3.083042   \n",
       "4          5.490820          5.715251          5.692676          1.804685   \n",
       "\n",
       "0  Dapma7bEVm629948  \n",
       "0          4.201193  \n",
       "1          4.118649  \n",
       "2          3.239676  \n",
       "3          3.390129  \n",
       "4          4.363281  \n",
       "\n",
       "[5 rows x 11883 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#save to CSV file\n",
    "control_df.to_csv('control_df.txt',sep='\\t', index=False)\n",
    "low_df.to_csv('low_df.txt',sep='\\t', index=False)\n",
    "high_df.to_csv('high_df.txt',sep='\\t', index=False)\n",
    "\n",
    "low_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "ab512aab-0cae-40f5-9bb5-d71c29ab3adb",
   "metadata": {},
   "outputs": [],
   "source": [
    "#get the order of genes to add to alphas file\n",
    "gene_order_control = control_df.columns.tolist()\n",
    "gene_order = gene_order_control[1:]\n",
    "gene_order_df = pd.DataFrame(gene_order,columns = ['Genes'])\n",
    "gene_order_df.to_csv('gene_order.csv')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
